New integration featuring the green planet energy prices API for tariff with dynamic prices

petschni opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->



 🌱 Add Green Planet Energy Integration


This PR introduces a new integration for **Green Planet Energy**, a German renewable energy provider, offering real-time electricity pricing data to help Home Assistant users optimize their power consumption based on hourly electricity rates.



- **29 Sensors Total:**
 - **24 Hourly Price Sensors**: `sensor.gpe_price_00` through `sensor.gpe_price_23` showing electricity prices for each hour
- **5 Statistical Sensors**: For energy optimization and monitoring:
  - `sensor.gpe_current_price` - Current hour's electricity price
  - `sensor.gpe_highest_price_today` - Peak price of the day
  - `sensor.gpe_lowest_price_day` - Cheapest price during day hours (06:00-18:00)
  - `sensor.gpe_lowest_price_night` - Cheapest price during night hours (18:00-06:00)
  - `sensor.gpe_price_chart_24h` - 24-hour chart data for visualization

The integration uses the following dependency: https://github.com/petschni/greenplanet-energy-api

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/40279
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] 

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. 

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository. (I chose small one since I am new to the project and don't think I have the overall knowledge to do really good review) (1) https://github.com/home-assistant/core/pull/149075#pullrequestreview-3089551863 (2) https://github.com/home-assistant/core/pull/149910#pullrequestreview-3089563750

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

petschni committed:
feat: add Green Planet Energy integration

- Add hourly electricity price sensors for business hours (9-18)
- Implement JSON-RPC API client with proper error handling
- Include comprehensive test suite with >95% coverage
- Meet Home Assistant Core Bronze quality standards
- Support config flow with unique entry validation
- Add runtime data for quality scale compliance

petschni committed:
Refactor Green Planet Energy integration

- Removed config flow, coordinator, and sensor files as part of the integration cleanup.
- Deleted constants and manifest files to streamline the integration structure.
- Updated the sensor implementation to support 24-hour pricing and added special sensors for highest, lowest, and current prices.
- Enhanced error handling and response processing in the API communication.
- Added installation script for easier setup of the integration.
- Updated tests to reflect changes in sensor setup and values, ensuring coverage for new features.
- Translated strings to English and improved naming conventions for clarity.

petschni committed:
feat: add support for lowest price day and night sensors in Green Planet Energy integration

petschni committed:
feat: add 24h price chart sensor and update data processing for tomorrow's prices

petschni committed:
fix: correct spelling in README and update function names in config flow for clarity

petschni committed:
feat: update Green Planet Energy integration to use new API methods and improve error handling

petschni committed:
Merge branch 'dev' into feat/greenplanet

petschni committed:
Merge branch 'dev' into feat/greenplanet

petschni committed:
Merge branch 'dev' into feat/greenplanet

petschni commented:
@homeassistant  did sign the cla now - can you recheck?

petschni committed:
Merge branch 'dev' into feat/greenplanet

joostlek commented on headers.txt:
Not sure why this is here

petschni replied:
left over - will delete it

joostlek commented on homeassistant/components/green_planet_energy/README.md:
Let's not have this readme here

petschni replied:
ok - deleted

joostlek commented on homeassistant/components/green_planet_energy/__init__.py:
```diff
@@ -0,0 +1,44 @@
+    # Set runtime data for quality scale compliance
+    entry.runtime_data = coordinator
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```
please use entry.runtime_data instead

joostlek commented on homeassistant/components/green_planet_energy/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+DATA_SCHEMA = vol.Schema({})
+
+
+async def prepare_config_entry(
+    hass: HomeAssistant, data: dict[str, Any]
+) -> dict[str, Any]:
+    """Prepare the config entry for Green Planet Energy.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    # For this integration, no API validation is required
+    # as the API might require authentication or not be publicly accessible
+    _LOGGER.info("Green Planet Energy integration setup - skipping API validation")
+    return {"title": "Green Planet Energy"}
```
This does nothing tho

petschni replied:
You are right - initially I thought it would require username and password as the input to access the API but that is not the case.

joostlek commented on homeassistant/components/green_planet_energy/const.py:
```diff
@@ -0,0 +1,19 @@
+
+DOMAIN = "green_planet_energy"
+
+# Config flow keys
+CONF_API_URL = "api_url"
+
+# Default values
+DEFAULT_API_URL = "https://mein.green-planet-energy.de/p2"
+DEFAULT_SCAN_INTERVAL = 60  # minutes
+
+# API constants
+API_METHOD = "getVerbrauchspreisUndWindsignal"
+AGGREGATS_ZEITRAUM = ""
+AGGREGATS_TYP = ""
+SOURCE = "Portal"
+
+# Sensor keys for each hour (0-23)
+SENSOR_HOURS = list(range(24))
```
I don't think we need all right?

joostlek commented on homeassistant/components/green_planet_energy/coordinator.py:
```diff
@@ -0,0 +1,62 @@
+        try:
+            return await self.api.get_electricity_prices()
+        except GreenPlanetEnergyConnectionError as err:
+            _LOGGER.warning(
+                "Connection error fetching data from Green Planet Energy API: %s", err
+            )
+            # Return empty data instead of raising an error
+            # This prevents the integration from failing completely
+            return {}
+        except GreenPlanetEnergyAPIError as err:
+            _LOGGER.error(
+                "API error fetching data from Green Planet Energy API: %s", err
+            )
+            # Return empty data instead of raising an error
+            # This prevents the integration from failing completely
+            return {}
```
if we couldn't get data we should raise UpdateFailed

joostlek commented on homeassistant/components/green_planet_energy/coordinator.py:
```diff
@@ -0,0 +1,62 @@
+            # This prevents the integration from failing completely
+            return {}
+
+    async def async_shutdown(self) -> None:
+        """Shutdown coordinator."""
+        await super().async_shutdown()
+        await self.api.close()
```
what does close do?

joostlek commented on homeassistant/components/green_planet_energy/quality_scale.yaml:
```diff
@@ -0,0 +1,24 @@
+rules:
```
Please add and evaluate all rules

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,366 @@
+    hour: int
+
+
+SENSOR_HOURLY_DESCRIPTIONS: list[GreenPlanetEnergySensorEntityDescription] = [
+    GreenPlanetEnergySensorEntityDescription(
+        key=f"gpe_price_{hour:02d}",
+        translation_key=f"price_{hour:02d}",
+        native_unit_of_measurement=f"{CURRENCY_EURO}/{UnitOfEnergy.KILO_WATT_HOUR}",
+        device_class=SensorDeviceClass.MONETARY,
+        suggested_display_precision=4,
+        hour=hour,
+    )
+    for hour in SENSOR_HOURS
+]
```
I am not sure how 24 entities are, wouldn't a service in the end be more user friendly?

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,366 @@
+        # Use fixed unique_id with just the key for predictable entity IDs
+        self._attr_unique_id = description.key
+
+        # Set explicit entity_id to ensure gpe_ prefix for all sensors
+        if description.hour >= 0:
+            # For hourly sensors, use gpe_price_XX format
+            self.entity_id = f"sensor.{description.key}"
+        else:
+            # For special sensors, use the full key
+            self.entity_id = f"sensor.{description.key}"
```
we should not change the entity_id

petschni replied:
But got obsolete with the new implementation anyway.

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,366 @@
+            # For special sensors, use the full key
+            self.entity_id = f"sensor.{description.key}"
+
+        # Set appropriate name based on sensor type
+        if description.hour >= 0:
+            self._attr_name = f"Price {description.hour:02d}:00"
+        else:
+            # For special sensors, the name will be set via translation_key
+            self._attr_name = None
```
we can also use translations for this

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,366 @@
+        if not self.coordinator.data:
+            return None
+
+        # Handle different sensor types
+        if self.entity_description.key == "gpe_highest_price_today":
+            return self._get_highest_price_today()
+        if self.entity_description.key == "gpe_lowest_price_day":
+            return self._get_lowest_price_day()
+        if self.entity_description.key == "gpe_lowest_price_night":
+            return self._get_lowest_price_night()
+        if self.entity_description.key == "gpe_current_price":
+            return self._get_current_price()
+        if self.entity_description.key == "gpe_price_chart_24h":
+            return self._get_current_price()  # Use current price as main value
```
extend the entity description to take a callable and move the entity specifics to the entity

petschni replied:
hope my change is what you meant - updated the code

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,366 @@
+
+        return chart_data
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return additional state attributes."""
+        if self.entity_description.hour >= 0:
+            # Regular hourly sensor
+            return {
+                "hour": self.entity_description.hour,
+                "time_slot": f"{self.entity_description.hour:02d}:00-{self.entity_description.hour + 1:02d}:00",
```
we should not put large objects in the state attributes, those are all saved in the database, so we should not

petschni replied:
changed

petschni committed:
Remove the `README.md` file and clean up the `__init__.py` by removing unused imports and redundant lines of code.

petschni committed:
remove unused config method

petschni committed:
added updateFailed method

petschni committed:
add quality scale informaiton for other levels

petschni committed:
reduce size of the price chart

petschni committed:
use translation

petschni committed:
introduced callable

petschni committed:
removing single enteties and providing a service for getting the price for a specific hour

petschni committed:
remove non requird constant

petschni committed:
Merge branch 'dev' into feat/greenplanet

petschni committed:
remove unused method

petschni commented:
@joostlek thanks for the review! I applied your change requests. Got a bit bigger due to the changes to a service but I guess it makes more sense with the service. Let me know if there is more.

joostlek commented on homeassistant/components/green_planet_energy/__init__.py:
```diff
@@ -0,0 +1,109 @@
+from .coordinator import GreenPlanetEnergyUpdateCoordinator
+from .sensor import get_current_price
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+# Config entry only (no YAML config)
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
+
+# Service schema
+SERVICE_GET_PRICE_SCHEMA = vol.Schema(
+    {
+        vol.Optional("hour"): vol.All(vol.Coerce(int), vol.Range(min=0, max=23)),
+    }
+)
```
Oh, maybe I should have added this, but we should do one platform per PR to make it still easy to review. So let's split the service part off for now, and add that in a subsequent PR

petschni committed:
remove service

petschni committed:
Merge branch 'dev' into feat/greenplanet

joostlek commented on homeassistant/components/green_planet_energy/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+# Config entry only (no YAML config)
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Green Planet Energy component."""
+    return True
```
```suggestion
```

joostlek commented on homeassistant/components/green_planet_energy/__init__.py:
```diff
@@ -0,0 +1,38 @@
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
Let's extend the ConfigEntry type as described in the runtime-data rule

petschni replied:
added a greenplanetconfigentry type

joostlek commented on homeassistant/components/green_planet_energy/config_flow.py:
```diff
@@ -0,0 +1,36 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        # Check if integration is already configured
+        await self.async_set_unique_id(DOMAIN)
+        self._abort_if_unique_id_configured()
```
Use the `single_config_entry` option in the manifest instead

joostlek commented on homeassistant/components/green_planet_energy/config_flow.py:
```diff
@@ -0,0 +1,36 @@
+        await self.async_set_unique_id(DOMAIN)
+        self._abort_if_unique_id_configured()
+
+        if user_input is not None:
+            return self.async_create_entry(title="Green Planet Energy", data=user_input)
```
Let's still try to fetch some data. Mostly because the config flow is the best way to tell the user something is wrong. Maybe they have DNS set up incorrectly, or they have a pi-hole running which blocks it. This way we can guarantee that it works.

petschni replied:
added a test

joostlek commented on homeassistant/components/green_planet_energy/coordinator.py:
```diff
@@ -0,0 +1,47 @@
+            update_interval=timedelta(minutes=DEFAULT_SCAN_INTERVAL),
+            config_entry=config_entry,
+        )
+        self.session = async_get_clientsession(hass)
+        self.api = GreenPlanetEnergyAPI(session=self.session)
```
```suggestion
        self.api = GreenPlanetEnergyAPI(session=async_get_clientsession(hass))
```

joostlek commented on homeassistant/components/green_planet_energy/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+
+  # Silver level - Enhanced functionality
+  entity-unavailable:
+    status: todo
+    comment: Integration needs to implement proper availability handling in coordinator.
+  parallel-updates:
+    status: todo
+    comment: Integration needs to set PARALLEL_UPDATES constant in sensor.py.
+  reauthentication-flow:
+    status: exempt
+    comment: Integration does not require authentication.
+  reconfiguration-flow:
+    status: exempt
+    comment: No reconfigurable parameters.
+  entity-category:
+    status: todo
+    comment: Integration could benefit from categorizing diagnostic entities.
+
+  # Gold level - Advanced features
+  devices:
+    status: todo
+    comment: Integration does not create device entries, only entities.
+  diagnostics:
+    status: todo
+    comment: Integration needs diagnostics.py implementation.
+  entity-translations: done
+  exception-translations:
+    status: exempt
+    comment: Integration uses simple service errors that don't require translation.
+  icon-translations:
+    status: exempt
+    comment: Integration uses static icons that don't change based on state.
+
+  # Platinum level - Highest quality standards
+  async-dependency:
+    status: exempt
+    comment: The greenplanet-energy-api library uses aiohttp and is fully async.
+  strict-typing:
+    status: todo
+    comment: Integration needs to be added to .strict-typing and implement typed ConfigEntry.
```
I think you used AI for this, please don't do that, it's easier to copy paste and be sure to give the rules itself a read to understand what it does :)

petschni replied:
yeah - actually did not find the template at first but did copy it now and checked the boxes

petschni replied:
Ok I guess I am a bit confused. I now used the following format

## Bronze
- [ ] `action-setup` - Service actions are registered in async_setup
- [x] `appropriate-polling` - If it's a polling integration, set an appropriate polling interval
- [ ] `brands` - Has branding assets available for the integration
- [ ] `common-modules` - Place common patterns in common modules
- [x] `config-flow-test-coverage` - Full test coverage for the config flow
- [x] `config-flow` - Integration needs to be able to be set up via the UI
    - [ ] Uses `data_description` to give context to fields
    - [x] Uses `ConfigEntry.data` and `ConfigEntry.options` correctly
- [x] `dependency-transparency` - Dependency transparency
- [x] `docs-actions` - The documentation describes the provided service actions that can be used
- [x] `docs-high-level-description` - The documentation includes a high-level description of the integration brand, product, or service
- [x] `docs-installation-instructions` - The documentation provides step-by-step installation instructions for the integration, including, if needed, prerequisites
- [x] `docs-removal-instructions` - The documentation provides removal instructions
- [ ] `entity-event-setup` - Entity events are subscribed in the correct lifecycle methods
- [x] `entity-unique-id` - Entities have a unique ID
- [x] `has-entity-name` - Entities use has_entity_name = True
- [x] `runtime-data` - Use ConfigEntry.runtime_data to store runtime data
- [x] `test-before-configure` - Test a connection in the config flow
- [x] `test-before-setup` - Check during integration initialization if we are able to set it up correctly
- [x] `unique-config-entry` - Don't allow the same device or service to be able to be set up twice

## Silver
- [ ] `action-exceptions` - Service actions raise exceptions when encountering failures
- [ ] `config-entry-unloading` - Support config entry unloading
- [ ] `docs-configuration-parameters` - The documentation describes all integration configuration options
- [ ] `docs-installation-parameters` - The documentation describes all integration installation parameters
- [ ] `entity-unavailable` - Mark entity unavailable if appropriate
- [ ] `integration-owner` - Has an integration owner
- [ ] `log-when-unavailable` - If internet/device/service is unavailable, log once when unavailable and once when back connected
- [ ] `parallel-updates` - Number of parallel updates is specified
- [ ] `reauthentication-flow` - Reauthentication needs to be available via the UI
- [ ] `test-coverage` - Above 95% test coverage for all integration modules

## Gold
- [ ] `devices` - The integration creates devices
- [ ] `diagnostics` - Implements diagnostics
- [ ] `discovery-update-info` - Integration uses discovery info to update network information
- [ ] `discovery` - Devices can be discovered
- [ ] `docs-data-update` - The documentation describes how data is updated
- [ ] `docs-examples` - The documentation provides automation examples the user can use.
- [ ] `docs-known-limitations` - The documentation describes known limitations of the integration (not to be confused with bugs)
- [ ] `docs-supported-devices` - The documentation describes known supported / unsupported devices
- [ ] `docs-supported-functions` - The documentation describes the supported functionality, including entities, and platforms
- [ ] `docs-troubleshooting` - The documentation provides troubleshooting information
- [ ] `docs-use-cases` - The documentation describes use cases to illustrate how this integration can be used
- [ ] `dynamic-devices` - Devices added after integration setup
- [ ] `entity-category` - Entities are assigned an appropriate EntityCategory
- [ ] `entity-device-class` - Entities use device classes where possible
- [ ] `entity-disabled-by-default` - Integration disables less popular (or noisy) entities
- [ ] `entity-translations` - Entities have translated names
- [ ] `exception-translations` - Exception messages are translatable
- [ ] `icon-translations` - Entities implement icon translations
- [ ] `reconfiguration-flow` - Integrations should have a reconfigure flow
- [ ] `repair-issues` - Repair issues and repair flows are used when user intervention is needed
- [ ] `stale-devices` - Stale devices are removed

## Platinum
- [ ] `async-dependency` - Dependency is async
- [ ] `inject-websession` - The integration dependency supports passing in a websession
- [ ] `strict-typing` - Strict typing


But then the "linter" complained and if I look at other integrations they are also using more the format I was using before.

petschni replied:
I think I understood now - hope it makes more sense now with the boxes ticked.

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,260 @@
+                "time_slot": f"{lowest_hour:02d}:00-{lowest_hour + 1:02d}:00"
+                if lowest_hour is not None
+                else None,
+                "period": "day (06:00-18:00)",
+            }
+
+        if self.entity_description.key == "gpe_lowest_price_night":
+            # Find the hour with the lowest price during night (18-6)
+            lowest_price = get_lowest_price_night(self.coordinator.data)
+            lowest_hour = None
+            if lowest_price is not None and self.coordinator.data:
+                # Check evening hours (18-23)
+                for hour in range(18, 24):
+                    price_key = f"gpe_price_{hour:02d}"
+                    if self.coordinator.data.get(price_key) == lowest_price:
+                        lowest_hour = hour
+                        break
+                # Check early morning hours (0-5) if not found
+                if lowest_hour is None:
+                    for hour in range(6):
+                        price_key = f"gpe_price_{hour:02d}"
+                        if self.coordinator.data.get(price_key) == lowest_price:
+                            lowest_hour = hour
+                            break
+            return {
+                "lowest_price_hour": lowest_hour,
+                "time_slot": f"{lowest_hour:02d}:00-{lowest_hour + 1:02d}:00"
+                if lowest_hour is not None
+                else None,
+                "period": "night (18:00-06:00)",
+            }
+
+        if self.entity_description.key == "gpe_current_price":
+            current_hour = dt_util.now().hour
+            return {
+                "current_hour": current_hour,
+                "time_slot": f"{current_hour:02d}:00-{current_hour + 1:02d}:00",
+            }
+
+        return None
```
Let's remove the extra state attributes for now?

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,260 @@
+        self.entity_description = description
+        # Use fixed unique_id with just the key for predictable entity IDs
+        self._attr_unique_id = description.key
+        # Set entity_id to use the full key
+        self.entity_id = f"sensor.{description.key}"
+        # Use translation_key for all sensor names
+        self._attr_name = None
```
```suggestion
```

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,260 @@
+    @property
+    def native_value(self) -> float | None:
+        """Return the state of the sensor."""
+        if not self.coordinator.data:
+            return None
+
+        # Use value_fn to get calculated values
```
```suggestion
```
coordinator data is never None

joostlek replied:
Or, if it could be an empty dict, it should raise `UpdateFailed`

petschni replied:
I removed the "none" case. I guess the updatefailed expection should be raised in the coordinator

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,260 @@
+        key="gpe_highest_price_today",
+        translation_key="highest_price_today",
+        native_unit_of_measurement=f"{CURRENCY_EURO}/{UnitOfEnergy.KILO_WATT_HOUR}",
+        device_class=SensorDeviceClass.MONETARY,
+        suggested_display_precision=4,
+        value_fn=get_highest_price_today,
+    ),
+    GreenPlanetEnergySensorEntityDescription(
+        key="gpe_lowest_price_day",
+        translation_key="lowest_price_day",
+        native_unit_of_measurement=f"{CURRENCY_EURO}/{UnitOfEnergy.KILO_WATT_HOUR}",
+        device_class=SensorDeviceClass.MONETARY,
+        suggested_display_precision=4,
+        value_fn=get_lowest_price_day,
+    ),
+    GreenPlanetEnergySensorEntityDescription(
+        key="gpe_lowest_price_night",
+        translation_key="lowest_price_night",
+        native_unit_of_measurement=f"{CURRENCY_EURO}/{UnitOfEnergy.KILO_WATT_HOUR}",
+        device_class=SensorDeviceClass.MONETARY,
+        suggested_display_precision=4,
+        value_fn=get_lowest_price_night,
+    ),
+    GreenPlanetEnergySensorEntityDescription(
+        key="gpe_current_price",
+        translation_key="current_price",
+        native_unit_of_measurement=f"{CURRENCY_EURO}/{UnitOfEnergy.KILO_WATT_HOUR}",
+        device_class=SensorDeviceClass.MONETARY,
```
Device class monetary is incorrect, that can only be used for a balance, as eur/kWh is not a balance you can't use it

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,260 @@
+            return None
+
+        # Use value_fn to get calculated values
+        if self.entity_description.value_fn:
+            return self.entity_description.value_fn(self.coordinator.data)
+
+        # Fallback to coordinator data
+        return self.coordinator.data.get(self.entity_description.key)
```
They all have value_fns

petschni replied:
removed the fallback - but still need the return or ?

petschni committed:
Update homeassistant/components/green_planet_energy/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

petschni committed:
Update homeassistant/components/green_planet_energy/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

petschni committed:
Update homeassistant/components/green_planet_energy/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

petschni committed:
updated quality scale and checked all levels

petschni committed:
remove device class

petschni committed:
remove fallback for names

petschni committed:
remove most extra attributes

petschni committed:
set single_config_entry

petschni committed:
test api connection during setup

petschni committed:
add custom config type

petschni committed:
removed non return case

petschni committed:
Merge branch 'dev' into feat/greenplanet

petschni committed:
Merge branch 'dev' into feat/greenplanet

petschni committed:
Merge branch 'dev' into feat/greenplanet

joostlek commented on homeassistant/components/green_planet_energy/config_flow.py:
```diff
@@ -0,0 +1,64 @@
+DATA_SCHEMA = vol.Schema({})
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect."""
+    session = async_get_clientsession(hass)
+    api = GreenPlanetEnergyAPI(session=session)
+
+    # Test the connection by trying to fetch data
+    await api.get_electricity_prices()
+
+    # Return info that you want to store in the config entry.
+    return {"title": "Green Planet Energy"}
+
+
+class GreenPlanetEnergyConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Green Planet Energy."""
+
+    VERSION = 1
+    MINOR_VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except GreenPlanetEnergyConnectionError:
+                errors["base"] = "cannot_connect"
+            except GreenPlanetEnergyAPIError:
+                errors["base"] = "invalid_auth"
+            except Exception:
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
```
I think we can inline this

joostlek commented on homeassistant/components/green_planet_energy/config_flow.py:
```diff
@@ -0,0 +1,64 @@
+                return self.async_create_entry(title=info["title"], data=user_input)
+
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors
```
can't you omit the schema here

joostlek commented on homeassistant/components/green_planet_energy/const.py:
```diff
@@ -0,0 +1,16 @@
+
+DOMAIN = "green_planet_energy"
+
+# Config flow keys
+CONF_API_URL = "api_url"
```
this is not used

joostlek commented on homeassistant/components/green_planet_energy/coordinator.py:
```diff
@@ -0,0 +1,46 @@
+class GreenPlanetEnergyUpdateCoordinator(DataUpdateCoordinator[dict[str, Any]]):
+    """Class to manage fetching data from Green Planet Energy API."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
```
```suggestion
    def __init__(self, hass: HomeAssistant, config_entry: GreenPlanetEnergyConfigEntry) -> None:
```
You might need to define the `type` in the coordinator file

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,239 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_highest_price_today(data: dict[str, Any]) -> float | None:
+    """Get the highest price from today's data."""
+    if not data:
+        return None
+
+    prices = []
+    for hour in range(24):
+        price_key = f"gpe_price_{hour:02d}"
+        if price_key in data:
+            price = data[price_key]
+            if price is not None:
+                prices.append(price)
+
+    return max(prices) if prices else None
+
+
+def get_lowest_price_day(data: dict[str, Any]) -> float | None:
+    """Get the lowest price during day hours (6-18)."""
+    if not data:
+        return None
+
+    prices = []
+    for hour in range(6, 18):  # Day period: 6:00 to 18:00
+        price_key = f"gpe_price_{hour:02d}"
+        if price_key in data:
+            price = data[price_key]
+            if price is not None:
```
Would it make sense to move these calculations to the package?

joostlek replied:
Yes but even then, ideally we don't have that much code that isn't directly HA specific here. So I would still opt to make this part ofthe lib

petschni replied:
@joostlek  I moved it to the lib now.

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,239 @@
+        # All our sensors have value_fn, so this should never happen
+        return None
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return additional state attributes."""
```
we should not put large data objects in the state attributes. Instead we could use action responses, but that's not for this PR

petschni replied:
ok

joostlek replied:
Why do we still have this then? :)

petschni replied:
ah - ok was just meant that I recognized the comment - not that I changed it - sorry.

petschni committed:
removed unused config option

petschni committed:
remove non required DATA_SCHEMA

petschni committed:
inline function

petschni committed:
adding a custom type

petschni commented:
@joostlek adjusted the code

petschni committed:
Merge branch 'dev' into feat/greenplanet

joostlek commented on homeassistant/components/green_planet_energy/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                session = async_get_clientsession(self.hass)
+                api = GreenPlanetEnergyAPI(session=session)
```
```suggestion
            session = async_get_clientsession(self.hass)
            api = GreenPlanetEnergyAPI(session=session)
            try:
```

joostlek commented on homeassistant/components/green_planet_energy/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+
+  # Platinum
+  async-dependency: todo
+  inject-websession: todo
```
```suggestion
  inject-websession: done
```

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,239 @@
+class GreenPlanetEnergySensorEntityDescription(SensorEntityDescription):
+    """Describes Green Planet Energy sensor entity."""
+
+    value_fn: Callable[[dict[str, Any]], float | None] | None = None
```
```suggestion
    value_fn: Callable[[dict[str, Any]], float | None]
```

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,239 @@
+    @property
+    def native_value(self) -> float | None:
+        """Return the state of the sensor."""
+        # Use value_fn to get calculated values
+        if self.entity_description.value_fn:
+            return self.entity_description.value_fn(self.coordinator.data)
+
+        # All our sensors have value_fn, so this should never happen
+        return None
```
```suggestion
        return self.entity_description.value_fn(self.coordinator.data)
```

joostlek commented on homeassistant/components/green_planet_energy/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+  test-coverage: done
+
+  # Gold
+  devices:
+    status: exempt
+    comment: Only creates sensors.
```
You can add a device_info with entry_type service to make the data appear grouped per device

petschni replied:
done

joostlek commented on homeassistant/components/green_planet_energy/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+  entity-device-class: todo
+  entity-disabled-by-default: todo
+  entity-translations: done
+  exception-translations: done
```
the exceptions raised in the coordinator are not translated yet

petschni replied:
added

joostlek commented on homeassistant/components/green_planet_energy/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+    comment: The integration registers no actions.
+  appropriate-polling: done
+  brands: done
+  common-modules:
+    status: exempt
+    comment: Not really potential for common modules.
```
you have coordinator.py

joostlek commented on homeassistant/components/green_planet_energy/strings.json:
```diff
@@ -0,0 +1,33 @@
+        "name": "Highest price today"
+      },
+      "lowest_price_day": {
+        "name": "Lowest price day (06:00-18:00)"
+      },
+      "lowest_price_night": {
+        "name": "Lowest price night (18:00-06:00)"
```
who decided what is day and what is night?

petschni replied:
Me ;) Basically the idea was that people would look for two timeslots. One during the day where you can run energy intensive things and one during the night. But I am totally open for a better definition or idea here. It was just what I came across when using the plugin that this proofed to be handy. 

The other idea would be to align this with sunset and sunrise but then it would have to be much more dynamic.

petschni replied:
done

joostlek commented on homeassistant/components/green_planet_energy/strings.json:
```diff
@@ -0,0 +1,33 @@
+      "unknown": "Unexpected error occurred"
+    },
+    "abort": {
+      "already_configured": "Service is already configured"
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]"
```

joostlek reviewed (changes requested):
Also, please send me a message on discord :)

petschni committed:
Update homeassistant/components/green_planet_energy/strings.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

petschni committed:
Update homeassistant/components/green_planet_energy/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

petschni committed:
Update homeassistant/components/green_planet_energy/quality_scale.yaml

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

petschni committed:
Update homeassistant/components/green_planet_energy/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

petschni committed:
Update homeassistant/components/green_planet_energy/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

petschni committed:
move logic to api package

petschni committed:
update api requirement

petschni committed:
Translate UpdateFailed exceptions in coordinator

- Add exception translations to strings.json
- Use translation_domain, translation_key, and translation_placeholders
- Follows Gold quality scale requirements for exception translations

petschni committed:
group as one service

petschni committed:
Merge branch 'dev' into feat/greenplanet

petschni committed:
Fix current_price sensor showing unknown

- Move current_price calculation from lambda to native_value property
- Ensures dt_util.now().hour is evaluated dynamically at each state update
- Fixes issue where hour was evaluated only once at initialization
- Current price now updates correctly based on the actual current hour

petschni committed:
Update API dependency to version 0.1.4

- Fixes missing price calculation methods in API package
- get_lowest_price_day now correctly returns today's day period prices (6-18h)
- Resolves 'unknown' state issue for lowest_price_day sensor
- API version 0.1.4 includes all required methods for price calculations

petschni committed:
Merge branch 'dev' into feat/greenplanet

joostlek commented on homeassistant/components/green_planet_energy/sensor.py:
```diff
@@ -0,0 +1,166 @@
+        if self.entity_description.key == "gpe_current_price":
+            return self.coordinator.api.get_current_price(
+                self.coordinator.data, dt_util.now().hour
+            )
```
I am not exactly sure why this needs special handling, it just looks fine to put in the entity description

petschni committed:
update tests

petschni committed:
update mock

petschni committed:
Merge branch 'dev' into feat/greenplanet

joostlek commented on tests/components/green_planet_energy/conftest.py:
```diff
@@ -0,0 +1,85 @@
+            "password": "test_password",
+            "customer_number": "12345",
+            "postal_code": "12345",
+        },
```
There's no data in the config entry though

petschni replied:
removed it

joostlek commented on tests/components/green_planet_energy/test_config_flow.py:
```diff
@@ -0,0 +1,117 @@
+    with patch(
+        "homeassistant.components.green_planet_energy.config_flow.GreenPlanetEnergyAPI"
+    ) as mock_api_class:
+        mock_api = AsyncMock()
```
We can update the conftest to also mock the config_flow module client

```py
@pytest.fixture
def mock_mealie_client() -> Generator[AsyncMock]:
    """Mock a Mealie client."""
    with (
        patch(
            "homeassistant.components.mealie.MealieClient",
            autospec=True,
        ) as mock_client,
        patch(
            "homeassistant.components.mealie.config_flow.MealieClient",
            new=mock_client,
        ),
    ):
        client = mock_client.return_value
        client.get_mealplans.return_value = MealplanResponse.from_json(
            load_fixture("get_mealplans.json", DOMAIN)
        )
        yield client
```

joostlek commented on tests/components/green_planet_energy/test_config_flow.py:
I am missing a test where we try to add the same entry twice and see that it aborts because there's only a single instance allowed

joostlek commented on tests/components/green_planet_energy/test_coordinator.py:
We don't test internals like this. As in, everything the coordinator does is reflected in entities, so we don't need to test this and can remove this file

joostlek commented on tests/components/green_planet_energy/test_sensor.py:
```diff
@@ -0,0 +1,84 @@
+            state.attributes.get("unit_of_measurement")
+            == f"{CURRENCY_EURO}/{UnitOfEnergy.KILO_WATT_HOUR}"
+        )
+
```
previous comment also eliminates the need for this test

petschni replied:
I think I am a bit lost which comment you are refering to here

petschni committed:
Use translation for sensor names with hardcoded time ranges

petschni committed:
remove unused config

petschni committed:
using test snapshots

petschni committed:
remove non required test

petschni commented on tests/components/green_planet_energy/test_sensor.py:
```diff
@@ -0,0 +1,84 @@
+
+    assert device is not None
+    assert device.name == "Green Planet Energy"
+    assert device.entry_type is dr.DeviceEntryType.SERVICE
```
done

petschni committed:
use fixtures

petschni committed:
set unique ID to domain

petschni committed:
move fixture

petschni committed:
add mock

petschni committed:
parameterized

petschni committed:
fixture

petschni committed:
add test

petschni committed:
remove special handling

petschni committed:
Merge branch 'dev' into feat/greenplanet

joostlek committed:
Merge branch 'dev' into feat/greenplanet

joostlek committed:
Fix

joostlek committed:
Fix
